paddle/fluid/framework/details/fuse_gradient_space_pass.cc

```diff
@@ -0,0 +1,154 @@
+//   Copyright (c) 2019 PaddlePaddle Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#include "paddle/fluid/framework/details/fuse_gradient_space_pass.h"
+#include <algorithm>
+#include <fstream>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "paddle/fluid/framework/ir/graph_helper.h"
+#include "paddle/fluid/framework/ir/node.h"
+#include "paddle/fluid/framework/op_info.h"
+#include "paddle/fluid/framework/op_registry.h"
+
+namespace paddle {
+namespace framework {
+namespace details {
+
+std::unique_ptr<ir::Graph> FuseGradientSpacePass::ApplyImpl(
+    std::unique_ptr<ir::Graph> graph) const {
+  ir::Graph& result = *graph;
+  graph->Set(kParamsAndGrads, new ParamsAndGrads);
+
+  std::unordered_map<std::string, ir::Node*> vars;
+  std::unordered_map<std::string, ir::Node*> ops;
+
+  // Get Vars and Ops
+  for (ir::Node* node : result.Nodes()) {
+    if (node->IsVar()) {
+      if (node->Var()) {
+        auto var_name = node->Var()->Name();
+        // Note: The graph may have the same name node. For example, parameter
+        // is the input of operator and it also is the output of optimizer;
+        vars.emplace(var_name, node);
+      }
+    } else {
+      GetTrainingGradVarName(node, &ops, &result);
+    }
+  }
+
+  // Set Gradients as Persistable
+  auto dtype = static_cast<proto::VarType::Type>(0);
+  auto& params_grads = result.Get<ParamsAndGrads>(kParamsAndGrads);
+  for (auto& p_g : params_grads) {
+    // Get gradient var
+    auto iter = vars.find(p_g.second);
+    PADDLE_ENFORCE(iter != vars.end());
+
+    // Set Persistable to prevent this var become reusable.
```

Comment:
Since grad vars are persistable, the allocation of memory space could be done at startup, no additional op in the main program needed then.

but still, need to know the memory pieces to do allreduce.

Reply:
I put the alloc_continuous_space_for_grad_op in `RunOnlyOnceProgram` which is runned in multi_device_pass, maybe it is better.